File: 1424-diagonal-traverse-ii/1424-diagonal-traverse-ii.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums) {
        int m = nums.size();
		vector<int> ans;
        queue<pair<int, int>> q;
        q.push({0,0}); // first row, first cell
		
		// BFS
        while (!q.empty()) {
            pair<int, int> p = q.front();
            q.pop();
            ans.push_back(nums[p.first][p.second]);
			
			// insert the element below, if in first column
            if (p.second == 0 && p.first+1 < m) q.push({p.first+1, p.second});
			
			// insert the right neighbour, if exists
            if (p.second+1 < nums[p.first].size())
                q.push({p.first, p.second+1});
        }
        return ans;
    }
};

File: 1424-diagonal-traverse-ii/1424-diagonal-traverse-ii.py
class Solution:
    def findDiagonalOrder(self, nums: List[List[int]]) -> List[int]:
        res = []
        
        for i in range(len(nums)):
            for j in range(len(nums[i])):
                if i + j >= len(res):
                    res.append([])
                res[i + j].append(nums[i][j])
        
        return [y for x in res for y in x[::-1]]